src/components/common/table/index.tsx

```diff
@@ -28,3 +32,94 @@ export default function Table<T extends TableRowType>(props: TableProps<T>): JSX
     ...(props.showPagination !== false ? { renderPaginationText } : {}),
   });
 }
+
+/**
+ * Ordered columns component that preserves reordered columns using callback functions.
+ * This table fetches column names from preferences and sets them if available.
+ * Also saves new columns order upon reordering, into user preferences (for the org scope).
+ */
+
+export type OrderPreserveableTableProps = {
+  setColumns: (columns: TableColumnType[]) => void;
+  id: string;
+};
+
+export function OrderPreserveableTable<T extends TableRowType>(
+  props: TableProps<T> & OrderPreserveableTableProps
+): JSX.Element {
+  const [initialized, setInitialized] = useState<boolean>(false);
+  const { setColumns, onReorderEnd, columns, id, ...tableProps } = props;
+  const { selectedOrganization, orgPreferences, reloadOrgPreferences } = useOrganization();
+
+  const getPreferenceName = useCallback(() => `${id}-columns`, [id]);
+
+  const getTableColumns = useCallback(
+    (columnNames: string[]): TableColumnType[] =>
+      columnNames
+        .map((columnName) => columns.find((column) => column.key === columnName))
+        .filter((column) => !!column) as TableColumnType[],
+    [columns]
+  );
+
+  const reorderHandler = async (reorderedColumns: string[]) => {
+    await PreferencesService.updateUserOrgPreferences(selectedOrganization.id, {
+      [getPreferenceName()]: reorderedColumns,
+    });
+    reloadOrgPreferences();
+    const columnsToSet = getTableColumns(reorderedColumns);
+    setColumns(columnsToSet);
+
+    if (onReorderEnd) {
+      onReorderEnd(reorderedColumns);
+    }
+  };
+
+  useEffect(() => {
+    const fetchSavedColumns = () => {
+      if (!orgPreferences) {
+        return;
+      }
+      setInitialized(true);
```

Comment:
Minor nitpick, but I wondered about setting initialized to true before finishing the initialization? It's unlikely that any of the subsequent steps would fail, though, since you programmed it defensively.

Reply:
Good call, for readability I can move this statement to the end after all the checks have completed. That is the correct logical order anyways.